### DISCUSSION TITLE
Introduce Fan speed percentage capability for MQTT fan integration

### DISCUSSION THREAD

**Discussion by jbouwh** [home-assistant/architecture#530] [General] [answered]
## Context

This as a follow up on architecture discussion https://github.com/home-assistant/architecture/issues/127.
The fan core identity fas been changed recently to support speed percentage. Using speeds (off, low, medium and high) is decrepated now. Since HA core-2021.3.2 the fan integration has been updated. 
[Separate fan speeds into percentages and presets modes #45407](https://github.com/home-assistant/core/pull/45407)
The mqtt.fan should follow this architecture update.

## Proposal

Update the mqtt fan integration to support setting a speed percentage.

## Consequences

This will enable to pass percentage based speeds to MQTT fans and te be conformant to the architecture changes of the `fan` integration and core identity updates.

The current 'speeds' attribute will remain working as it is now to support backwards compatiblity. Functions to translate between predefined speeds and percentages have been built into ha core already (`speed_to_percentage()` and `percentage_to_speed()` in `components/fan`)

If both `speeds` and `percentage` methods are advertised (or set through config) the predefined functions will be used to translate a percentage to predefined speed and back (`speed_to_percentage()` and `percentage_to_speed()` in `components/fan`).

A MQTT fan advertises a `percentage_command_topic` to ha core mqtt auto discovery it supports percentage based variable speeds.
A `percentage_state_topic` is used to report back the actual percentage set by the fan.
A `percentage_value_template` defines te template to extract the percentage from the payload.

**Comment by jbouwh** [comment]
In addition to address this as an issue. Using the mqtt fan integration using `set_speed` (the only available method now) results to warnings in the logging:

```
Logger: homeassistant.components.fan
Source: components/fan/__init__.py:231
Integration: Fan (documentation, issues)
First occurred: 7:30:00 (1 occurrences)
Last logged: 7:30:00

fan.set_speed is deprecated, use fan.set_percentage or fan.set_preset_mode instead.
```

**Reply by gmarteleur** [reply]
Thank you very much. It is just what I was expecting to be able to manage more speed options on my Fan. I use MQTT Discovery feature with ESP8266 system, but adding "percentage_command_topic" seems not working. Using the service "Fan: Set speed percentage", I receive new messages on  "speed_command_topic", but "percentage_command_topic" remains black. I'm using core version: core-2021.3.4
Are these features already implemented or it is an on going project?

**Reply by jbouwh** [reply]
> Thank you very much. It is just what I was expecting to be able to manage more speed options on my Fan. I use MQTT Discovery feature with ESP8266 system, but adding "percentage_command_topic" seems not working. Using the service "Fan: Set speed percentage", I receive new messages on "speed_command_topic", but "percentage_command_topic" remains black. I'm using core version: core-2021.3.4
> Are these features already implemented or it is an on going project?

Actually i am working on the implementation of the code, so please have sime patience.

**Reply by gmarteleur** [reply]
Ok. Thanks for your answer!!

**Comment by frenck** [comment]
Transferring the issue into a discussion.

**Comment by frenck** [comment]
I don't think this needs an architectural discussion? The architectural discussion on the whole fan model has been finished and implemented. This would be just implementing it additionally, right?

Unless I miss something; I would say: Go ahead and open a PR.

**Reply by emontnemery** [reply]
You're right, there's no need for an architecture discussion. A PR would be most welcome.

**Comment by jbouwh** [comment, marked as answer]
Ik could give it a try, give me some time :-). I have some experience with AppDaemon automation with hacks, but never did a PR for HA.

**Reply by jbouwh** [reply]
Link to the PR
https://github.com/home-assistant/core/pull/47944

**Comment by dwmw2** [comment]
It looks like ESPHome was still reporting and accepting raw speed values (e.g. 1-4) in the `percentage_state_topic` and `percentage_command_topic` respectively. I have a PR to fix that: https://github.com/esphome/esphome/pull/12027

I don't actually *use* HA (which is why I care about ESPHome getting the MQTT wrong; most HA users would be using the API anyway). Can someone confirm that this is the correct thing to do for HA and that HA really does behave as described in https://github.com/home-assistant/core/pull/47944
